Check for duplicate literal switch cases

ffried opened the pull request:
<!-- What's in this pull request? -->
This adds a check for duplicate literal switch cases. It covers string, integer and float literals. Corresponding tests were added as well.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
According to @CodaFi this should resolve [rdar://28301984](rdar://problem/28301984).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

ffried committed:
Check for duplicate literal switch cases

This adds a check for duplicate literal switch cases. It covers string, integer and float literals. Corresponding tests were added as well.

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -914,9 +948,59 @@ namespace {
+        // FIXME: Pessimistically using IEEEquad here is bad and we should
+        // actually figure out the bitwidth.  But it's too early in Sema.
+        auto *FLE = cast<FloatLiteralExpr>(EL);
+        decltype(FloatLiteralCache)::iterator it;
```
This was leftover from an experiment, could you remove it?

CodaFi commented on test/Sema/exhaustive_switch.swift:
```diff
@@ -558,3 +558,58 @@ func infinitelySized() -> Bool {
+     case 2.5: break // expected-note {{first occurence of identical literal pattern is here}}
+     case 2.5: break // expected-warning {{literal value is already handled by previous pattern; is this intentional?}}
+     case 3.5: break
+     default: break
```
These tests look good, but could you add some extra duplicate cases in

- Different positions e.g.

```swift
      case 1.5: break
      case 2.5: break //
      case 2.5: break //
      case 3.5: break
      case 2.5: break //
      default: break
```

- Different levels of normalization:

```swift
      case 1.5: break
      case 1.5000: break
      case 1.500: break

///
      case 1
      case 0b1
      case 0x1
```

CodaFi commented:
Xi should also take a look at this, but in the mean time 

@swift-ci please smoke test.

xwu commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3773,6 +3773,11 @@ NOTE(missing_particular_case,none,
+        "literal value is already handled by previous pattern; "
+        "is this intentional?",())
+NOTE(redundant_particular_literal_case_here,none,
+     "first occurence of identical literal pattern is here", ())
```
Spelling: this should be "occurrence"

xwu commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3773,6 +3773,11 @@ NOTE(missing_particular_case,none,
         "case is already handled by previous patterns; consider removing it",())
+WARNING(redundant_particular_literal_case,none,
+        "literal value is already handled by previous pattern; "
+        "is this intentional?",())
```
Nit: since there's no way to silence this warning without removing the redundant case, it's not actually possible for the user to answer the question "is this intentional?"--therefore, probably best to use the same wording as above (i.e. "consider removing it").

ffried commented:
@xwu done and done

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

jrose-apple commented:
I don't mind if this patch doesn't catch this case, but can you add a test like the following?

```swift
switch (x, y) {
case (0, 0): break
case (0, 1): break // not duplicate
case (1, 0): break // not duplicate
case (0, 0): break // duplicate
default: break
}
```

(and make sure it *doesn't* warn on the second and third cases)

nkcsgexi commented:
Could you also add a test for matching several cases in single case statement like:
```
func main(_ a: Int) -> String {
  switch a {
  case 1, 2: return ""
  case 1, 2: return ""
  case 2, 3: return ""
  default: return ""
  }
}
```

nkcsgexi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -914,9 +948,58 @@ namespace {
     SpaceEngine(TypeChecker &C, SwitchStmt *SS) : TC(C), Switch(SS) {}
 
+    bool checkRedundantLiteral(const Pattern *Pat, Expr *&PrevPattern) {
+      if (Pat->getKind() != PatternKind::Expr) return false;
```
nitpick: could we put `return false` into the next line?

nkcsgexi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -914,9 +948,58 @@ namespace {
+      if (Pat->getKind() != PatternKind::Expr) return false;
+      auto *ExprPat = cast<ExprPattern>(Pat);
+      auto *MatchExpr = ExprPat->getSubExpr();
+      if (!MatchExpr || !isa<LiteralExpr>(MatchExpr)) return false;
```
Same here.

ffried commented:
@jrose-apple Done. Added a test with a tuple with each of the literal types.

ffried commented:
@nkcsgexi Done: Added test with multiple cases handled in a single case statement for each of the literal type.

Also done: Added a line break, as well as curly brackets to the two ifs.

CodaFi commented:
@swift-ci please smoke test and merge

CodaFi commented:
@swift-ci please smoke test and merge

ianpartridge commented:
With this PR, the following code prints warnings:

```swift
let i: Int = 1
switch i {
  case 0: print("0")
  case 1, -1: print("1 or -1")  // Warning: Literal value is already handled by previous pattern; consider removing it
  default: print("something else")
}
```

That doesn't look right to me...?

ffried commented:
@ianpartridge Yep, that's definitively a bug. Will open another PR fixing this asap.

@CodaFi Should we revert this PR until then?

ffried commented:
@ianpartridge #11286 should fix this.

ianpartridge commented:
I confirm it does - thanks.
